share/eds/p2p/client.go

```diff
@@ -0,0 +1,143 @@
+package p2p
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	p2p_pb "github.com/celestiaorg/celestia-node/share/eds/p2p/pb"
+
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var log = logging.Logger("shrex/eds")
+
+// TODO(@distractedm1nd): make version suffix configurable
+var protocolID = protocol.ID("/shrex/eds/0.0.1")
+
+// Client is responsible for requesting EDSs for blocksync over the ShrEx/EDS protocol.
+// This client is run by light nodes and full nodes. For more information, see ADR #13
+type Client struct {
+	protocolID protocol.ID
+	host       host.Host
+}
+
+// NewClient creates a new ShrEx/EDS client.
+func NewClient(host host.Host) *Client {
+	return &Client{
+		host:       host,
+		protocolID: protocolID,
+	}
+}
+
+// RequestEDS requests the full EDS from one of the given peers.
+//
+// The peers are requested in a round-robin manner with retries until one of them gives a valid
+// response. Blocks forever until the context is canceled or a valid response is given.
+func (c *Client) RequestEDS(
+	ctx context.Context,
+	root share.Root,
+	peers peer.IDSlice,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	req := &p2p_pb.EDSRequest{Hash: root.Hash()}
+
+	// requests are retried for every peer until a valid response is received
+	edsCh := make(chan *rsmt2d.ExtendedDataSquare)
+	go func() {
+		for {
+			for _, to := range peers {
+				eds, err := c.doRequest(ctx, req, root, to)
+				if err != nil {
+					// TODO: should we exclude the peer from retries if we get an error?
+					log.Errorw("client: eds request to peer failed", "peer", to, "hash", root.String())
+				}
+
+				edsCh <- eds
+			}
+		}
```

Comment:
You should exit from this routine once the correct eds is received and cancel all other requests.

Reply:
done, but can you check if I did it correctly?